src/lib/sensor.rs

```diff
@@ -0,0 +1,116 @@
+use eyeball::{shared::Observable, Subscriber};
+use std::time::{Duration, SystemTime};
+
+pub trait SensESPSensor {
+    fn tick(&mut self);
+}
+
+pub trait Attachable<T> {
+    fn attach(&mut self) -> Subscriber<T>;
+}
+
+pub struct ConstantSensor<T> {
+    observable: Observable<T>,
+    duration: Duration,
+    last_measurement: SystemTime,
+    value: T,
+}
+
+impl<T: Copy> ConstantSensor<T> {
+    pub fn new(val: T, duration: Duration) -> Self {
+        let observable = Observable::new(val);
+        let last_measurement = SystemTime::now();
+        ConstantSensor::<T> {
+            observable,
+            value: val,
+            duration,
+            last_measurement,
+        }
+    }
+}
+
+impl<T: Copy> SensESPSensor for ConstantSensor<T> {
+    fn tick(&mut self) {
+        let now = SystemTime::now();
+
+        match now.duration_since(self.last_measurement) {
+            Ok(d) => match d >= self.duration {
+                true => {
+                    self.observable.set(self.value);
+                    self.last_measurement = now;
+                    ()
+                }
+                false => (),
+            },
+            Err(e) => log::error!("System time error on SensESP-rs tick: {:?}", e),
+        }
+    }
+}
+
+impl<T: Copy> Attachable<T> for ConstantSensor<T> {
+    fn attach(&mut self) -> Subscriber<T> {
+        self.observable.subscribe()
+    }
+}
+
+pub struct TimedSensor<T, F>
+where
+    T: Copy,
```

Comment:
This is an example of a typical generic constraint, `T: Copy` requires that the generic T implements the copy trait.  At compile time, a concrete implementation of `TimedSensor` is created for the exact type of `T`, for example, `i32` in the test example.  If it has a `Vec<T>`, at runtime it would be a concrete `Vec<i32>`